### PR TITLE
Fix some warnings in tests

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
Fix some of the warnings in the tests in order to restore the `-Werror` flag later.
- Remove test for `state_number_qobj`: (function replaced by `basis`.)
- scipy.sparse matrix operation replaced by `data.CSR` operation.
- Fix passing generator to `hstack`.
- Add some `filterwarnings` for `mcsolve` fallback to `sesolve` and coefficient parsing warning.

**Related issues or PRs**
TODO in #1850

**Changelog**
Fix some warnings in tests.

### CODE REVIEW

File: qutip/tests/solve/test_mcsolve.py [hunk 1]
```diff
@@ -272,6 +276,7 @@ def _regression_490_f2(t, args):
     return -t
 
 
+@pytest.mark.filterwarnings("ignore::UserWarning")
```
Comment (hodgestar) [inline qutip/tests/solve/test_mcsolve.py:279]:
```suggestion
@pytest.mark.filterwarnings("ignore:No c_ops, using sesolve:UserWarning")
```

File: qutip/random_objects.py [hunk 2]
```diff
@@ -178,10 +178,15 @@ def _rand_herm_sparse(N, density, pos_def):
     M = sp.coo_matrix((data, (row_idx, col_idx)),
                       dtype=complex, shape=(N, N))
     M = 0.5 * (M + M.conj().transpose())
-    if pos_def:
-        M.setdiag(np.abs(M.diagonal()) + np.sqrt(2) * N)
     M.sort_indices()
-    return _data.create(M)
+    rand_mat = _data.create(M)
```
[the remark this answers is not in the document]
Comment (Ericgig) [reply qutip/random_objects.py:182]: `setdiag`  was raising an sparse efficiency warning with `scipy.csr` matrix saying we should use `coo` or `lil` sparse format instead.

File: qutip/tests/core/test_coefficient.py [hunk 3]
```diff
@@ -263,6 +263,9 @@ def test_CoeffOptions():
     pytest.param("t + (0 if not 'something' else 1)", {},
                  lambda t: t + 1, id="branch")
 ])
+@pytest.mark.filterwarnings(
```
[the remark this answers is not in the document]
Comment (Ericgig) [reply qutip/tests/core/test_coefficient.py:266]: It's the string `'key'` that cause an issue. It's not smart enough to detect constant that are supposed to be string in the code string. The _right way_ to do it would be to pass to string as an args:
`coefficient("cos(t*dictionary[key])", args={'dictionary': {'key': 1}, 'key':'key'}` work without warning.

I think we should keep the fallback anyway, we don't want to remake a full parsing code and there will always be hedge case that we did not think about.

[the remark this answers is not in the document]
Comment (hodgestar) [reply qutip/tests/core/test_coefficient.py:266]: Ah, I see -- we are allowing bare words inside the string.

I would be in favor of deprecating that for version 5 (i.e. to make `t * dictionary["key"]` required if one wants a constant. That's simpler and consistent with Python. However, that's not a discussion for this PR, I don't think.

**Review by hodgestar — APPROVED** [review]
